### DISCUSSION TITLE
How to conditionally derive state base on the previous value

### DISCUSSION THREAD

**Discussion by moudy** [pmndrs/jotai#1304] [Q&A] [answered]
I was looking for a way to conditionally derive state based on the previous value. I thought `selectAtom` might provide the previous value but it doesn't. 

Here's a simple example of what I was looking to do with a hypothetical `reduceAtom` function that acts like `selectAtom` but also provides the previous value:

```js
const numberAtom = atom(1)
// reduceAtom doesn't exists
const lastEvenAtom = reduceAtom(numberAtom,  (next, previous) => next % 2 === 0 ? next : previous)

get(lastEvenAtom) // undefined

set(numberAtom, 2)
get(lastEvenAtom) // 2

set(numberAtom, 3)
get(lastEvenAtom) // still 2
```

Is there another way to accomplish this that I'm missing?

**Comment by dai-shi** [comment, marked as answer]
Something like this would work:
```js
const reduceAtom = (baseAtom, reducer) => {
  const derivedAtom = atom((get) => {
    const prev = get(derivedAtom) // TS can't infer type.
    const next = get(baseAtom)
    return reducer(prev, next) // parameters are swapped.
  })
  return derivedAtom
}
```

**Reply by moudy** [reply]
Oh interesting, thanks!

**Reply by dlindenkreuz** [reply]
Thanks! This was helpful for implementing a `latestValueAtom` that would return the latest value of two atoms.

```ts
import {shallow} from "zustand/shallow"

const latestValueAtom = <T>(
  atomA: Atom<T>,
  atomB: Atom<T>,
  compare: (lhs: T, rhs: T) => boolean = shallow
) => {
  const withPrev: Atom<readonly [T, T, T]> = atom((get) => {
    const [prevA, prevB, prevResult] = get(withPrev)
    const a = get(atomA)
    const b = get(atomB)
    const result = compare(prevA, a) ? (compare(prevB, b) ? prevResult : b) : a
    return [a, b, result] as const
  })
  // HACK: use private `init` param to get reducer functionality
  ;(withPrev as Atom<readonly [T, T, T]> & { init: never[] }).init = []
  // get only latest value
  return atom((get) => get(withPrev)[2])
}
```

Demo: https://codesandbox.io/s/musing-roentgen-p523hn?file=/src/App.tsx

**Reply by dai-shi** [reply]
right, my original answer doesn't work without `.init` hack. good catch.

@dlindenkreuz Can you make `latestValueAtom` with `selectAtom` without the explicit hack?

**Reply by dai-shi** [reply]
Maybe, selectAtom's reducer should receive prevSlice.

**Reply by dai-shi** [reply]
You can try this: https://github.com/pmndrs/jotai/pull/1738

**Reply by dlindenkreuz** [reply]
Nice, `prevSlice` is a really useful addition! 🙏

The new API made TypeScript reveal an edge case: When the atoms are initialized with different values, one of them needs to take precedence. I chose `atomB` which intuitively reflects the precedence behavior of JS object spreads, e.g. `{ ...a, ...b }`.

Check the updated version below:

```ts
import {shallow} from "zustand/shallow"

const latestValueAtom = <T>(
  atomA: Atom<T>,
  atomB: Atom<T>,
  compare: (lhs?: T, rhs?: T) => boolean = shallow
) => {
  const values = atom((get) => [get(atomA), get(atomB)] as const)
  const derived = selectAtom<readonly [T, T], readonly [T, T, T]>(
    values,
    // b takes precedence
    ([a, b], [prevA, prevB, prevResult] = [a, b, b]) => {
      const result = compare(prevA, a)
        ? compare(prevB, b)
          ? prevResult
          : b
        : a
      return [a, b, result] as const
    }
  )
  // get only result
  return atom((get) => get(derived)[2])
}
```

Demo: https://codesandbox.io/s/gallant-david-gwicmy?file=/src/App.tsx

An alternative solution could be to pass a tie-breaker instead of `compare` (not shown here):
`selectValue: (prevA: T | undefined, prevB: T | undefined, a: T, b: T) => T`

The tie-breaker could also have a different return type, e.g. allowing to merge a and b with their previous values:
```ts
// not tested, your mileage may vary
const defaultSelectValue = <T, R = T>(prevA: T | undefined, prevB: T | undefined, a: T, b: T) =>
  compare(prevA, a)
        ? compare(prevB, b)
          ? prevResult
          : b
        : a
const latestValueAtom = <T, R = T>(
  atomA: Atom<T>,
  atomB: Atom<T>,
  selectValue: (prevA: T | undefined, prevB: T | undefined, a: T, b: T) => R = defaultSelectValue<T, R>
) => // ...
```

**Reply by dai-shi** [reply]
looks nice. Yeah, I think #1738 is reasonable.
